gitpod-io/gitpod
# Implement project configurator

Initial pass to fix https://github.com/gitpod-io/gitpod/issues/4648

Depends on & contains https://github.com/gitpod-io/gitpod/pull/4846

What it does:

- Adds a `config` (text) field to `d_b_project`, so that Projects can have a custom `.gitpod.yml` content
- Adds a Project Configurator view in the dashboard, as final step in the Add Project flow (also currently accessible via a Project's UI)
- When determining the config of a new workspace, use this priority order:
    1. From repository
    2. (_new_) From Project DB
    3. From definitely-gp
    4. Derived (Go) or default
- When starting from a custom `.gitpod.yml` (from the Project DB), the initializer also places the uncomitted `.gitpod.yml` into the workspace (to encourage repo-based configs)

How to test:

- Create a new team
- Add a new project (either by inserting into `d_b_project`, or we set up a GitHub App integration -- @AlexTugarev could you help with the latter?)
- If the project already has a repo-based config, the configurator should indicate that and be disabled. If there is no repo config, the configurator should allow you to write a new `.gitpod.yml` (e.g. based on one of the presets) and then test it / iterate on it in a quick loop (with the build output on the right)

The UX of this is still rough, but the goal is to add this final step rather soon, and then iterate on the entire Add Project flow.
